pkg/kubectl/cmd/rollout/rollout_undo.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rollout
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/util/errors"
+	"k8s.io/kubernetes/pkg/util/wait"
+
+	"github.com/spf13/cobra"
+)
+
+// UndoOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type UndoOptions struct {
+	Filenames []string
+}
+
+const (
+	undo_long    = `undo rolls back to a previous rollout.`
+	undo_example = `# Rollback to the previous deployment
+$ kubectl rollout undo deployment/abc`
+)
+
+func NewCmdRolloutUndo(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &UndoOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "undo (TYPE NAME | TYPE/NAME) [flags]",
+		Short:   "undoes a previous rollout",
+		Long:    undo_long,
+		Example: undo_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(RunUndo(f, cmd, out, args, options))
+		},
+	}
+
+	cmd.Flags().Int64("to-revision", 0, "The revision to rollback to. Default to 0 (last revision).")
+	usage := "Filename, directory, or URL to a file identifying the resource to get from a server."
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	return cmd
+}
+
+func RunUndo(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string, options *UndoOptions) error {
+	if len(args) == 0 && len(options.Filenames) == 0 {
+		return cmdutil.UsageError(cmd, "Required resource not specified.")
+	}
+
+	toRevision := cmdutil.GetFlagInt64(cmd, "to-revision")
+	mapper, typer := f.Object()
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	r := resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest().
+		Flatten().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+	info := infos[0]
+	mapping := info.ResourceMapping()
+	rollbacker, err := f.Rollbacker(mapping)
```

Comment:
I don't like that we have two different calls to the factory, one just to filter out the invalid types, and one that actually does what we want. Please try to do both the filtering and returning of an event searcher in one call. The factory is already bloated.